crypto/crypto/src/multi_ed25519.rs

```diff
@@ -0,0 +1,547 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module provides an API for the accountable threshold multi-sig PureEdDSA signature scheme
+//! over the ed25519 twisted Edwards curve as defined in [RFC8032](https://tools.ietf.org/html/rfc8032).
+//!
+//! Signature verification also checks and rejects non-canonical signatures.
+
+use crate::ed25519::{
+    Ed25519PrivateKey, Ed25519PublicKey, Ed25519Signature, ED25519_PRIVATE_KEY_LENGTH,
+    ED25519_PUBLIC_KEY_LENGTH, ED25519_SIGNATURE_LENGTH,
+};
+use crate::{traits::*, HashValue};
+use anyhow::{anyhow, Result};
+use bit_vec::BitVec;
+use core::convert::TryFrom;
+use libra_crypto_derive::{DeserializeKey, SerializeKey, SilentDebug, SilentDisplay};
+use once_cell::sync::Lazy;
+use std::fmt;
+
+const MAX_NUM_OF_KEYS: usize = 32;
+const BITVEC_NUM_OF_BYTES: usize = MAX_NUM_OF_KEYS / 8;
+static BITVEC_FIRST_KEY_SET: Lazy<BitVec> =
+    Lazy::new(|| BitVec::from_bytes(&[0b1000_0000, 0b0000_0000, 0b0000_0000, 0b0000_0000]));
+
+/// Vector of private keys in the multi-key Ed25519 structure along with the threshold.
+#[derive(DeserializeKey, SilentDisplay, SilentDebug, SerializeKey)]
+pub struct MultiEd25519PrivateKey(Vec<Ed25519PrivateKey>, u8);
```

Comment:
what's the advantage of having 32 versus 255? or even 65K?

Reply:
Good question, reasoning:
- Bitcoin offers up to N=15 in P2SH and N=20 with raw keys. There are good reasons to support more than that and agreed that 32 meets most of the known custodial and escrow requirements.
- 32 keys implies 32 sigs, thus in the worst case we'll have 32*32 bytes for keys + 64*32 for sigs + 4 bytes for the bitvec + 1 byte for the key threshold. All in all, about 3K. If we supported 256 signatures we would need up to 24K per transaction. Obviously 65000 keys is too big for any blockchain. We've agreed we can tolerate up to 2-3K, so I picked the max.
- DoS, imagine a 1000 out-of 1000 where the last signature fails. The server spent time on 999 signatures only to realize it's invalid (and because you don't pay gas if the sig fails, it's cost free).
- With batch verification (which is 2.2x faster than iterative single verification), 32 sigs is analogous to individually checking 14-15 sigs. We've agreed that this is tolerable for DoS purposes.

Reply:
for context I've looked at several applications that provide multisig, worked with organizations that were using/looking to adopt multisig, and I've never seen them use something more complex than 2 out of 3 sigs. So in practice 32 should be more than enough.

Reply:
> DoS, imagine a 1000 out-of 1000 where the last signature fails. The server spent time on 999 signatures only to realize it's invalid (and because you don't pay gas if the sig fails, it's cost free).

I hadn't think about that, but isn't this becoming an issue then? A good way to make validators compute stuff would be to send an invalid 32 multisig tx no?

Reply:
@mimoo indeed, but the same attack could be achieved by submitting 14 invalid single-key transactions anyway. As mentioned above, a 32-keys sig will be equivalent to verifying 14-15 single sigs. Obviously we had to combine this with IP filtering and DoS protection offered by firewall rules and we had to agree on a number. For instance, if this limit was 100,000 then even with one tx request one could cause issues; 32 seems to be just tolerable and it also meets the criteria for most of the aforementioned requirements.